File: src/tools/projects/setActive.ts
import type { z } from "zod";
import type { Context, Tool } from "../types";
import { ProjectSetActiveSchema } from "../../schema/tool-inputs";

const schema = ProjectSetActiveSchema;

type Params = z.infer<typeof schema>;

export const setActiveHandler = async (context: Context, params: Params) => {
	const { projectId } = params;

	await context.cache.set("projectId", projectId.toString());

	return {
		content: [{ type: "text", text: `Switched to project ${projectId}` }],
	};
};

const tool = (): Tool<typeof schema> => ({
	name: "project-set-active",
	description: `
        - Use this tool to set the active project.
    `,
	schema,
	handler: setActiveHandler,
});

export default tool;

File: python/tools/projects/set_active.py
from schema.tool_inputs import ProjectSetActiveSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def set_active_project_handler(context: Context, params: ProjectSetActiveSchema) -> ToolResult:
    project_id = str(params.projectId)

    await context.cache.set("project_id", project_id)

    return ToolResult(content=f"Switched to project {project_id}")


def set_active_project_tool() -> Tool[ProjectSetActiveSchema]:
    definition = get_tool_definition("project-set-active")

    return Tool(
        name="project-set-active",
        description=definition["description"],
        schema=ProjectSetActiveSchema,
        handler=set_active_project_handler,
    )
